File: NGLess/Dependencies/embedded.c
#ifndef NO_EMBED_SAMTOOLS_BWA
#include "samtools_data.c"
#include "prodigal_data.c"
#include "bwa_data.c"
#include "megahit_data.c"
#include "minimap2_data.c"

const unsigned char* get_samtools_data () { return ngless_samtools_static; }
unsigned int get_samtools_len () { return ngless_samtools_static_len; }

const unsigned char* get_prodigal_data () { return ngless_prodigal_static; }
unsigned int get_prodigal_len () { return ngless_prodigal_static_len; }

const unsigned char* get_bwa_data () { return ngless_bwa_static; }
unsigned int get_bwa_len () { return ngless_bwa_static_len; }

const unsigned char* get_megahit_data() { return megahit_packaged_tar_gz; }
unsigned int get_megahit_len() { return megahit_packaged_tar_gz_len; }

const unsigned char* get_minimap2_data() { return ngless_minimap2_static; }
unsigned int get_minimap2_len() { return ngless_minimap2_static_len; }

#else

const unsigned char* get_samtools_data () { return ""; }
unsigned int get_samtools_len () { return 0; }

const unsigned char* get_prodigal_data () { return ""; }
unsigned int get_prodigal_len () { return 0; }

const unsigned char* get_bwa_data () { return ""; }
unsigned int get_bwa_len () { return 0; }

const unsigned char* get_megahit_data() { return ""; }
unsigned int get_megahit_len() { return 0; }

const unsigned char* get_minimap2_data() { return ""; }
unsigned int get_minimap2_len() { return 0; }

#endif

File: NGLess/Dependencies/Embedded.hs
{- Copyright 2015-2018 NGLess Authors
 - License: MIT
 -}
module Dependencies.Embedded
    ( samtoolsData
    , bwaData
    , prodigalData
    , megahitData
    , minimap2Data
    ) where

import qualified Data.ByteString as B
import qualified Data.ByteString.Unsafe as B

import Foreign.C.Types
import Foreign.C.String
import Data.Convertible

foreign import ccall safe "get_samtools_len" c_get_samtools_len :: CUInt
foreign import ccall safe "get_samtools_data" c_get_samtools_data :: CString

foreign import ccall safe "get_bwa_len" c_get_bwa_len :: CUInt
foreign import ccall safe "get_bwa_data" c_get_bwa_data :: CString

foreign import ccall safe "get_prodigal_len" c_get_prodigal_len :: CUInt
foreign import ccall safe "get_prodigal_data" c_get_prodigal_data :: CString

foreign import ccall safe "get_megahit_len" c_get_megahit_len :: CUInt
foreign import ccall safe "get_megahit_data" c_get_megahit_data :: CString

foreign import ccall safe "get_minimap2_len" c_get_minimap2_len :: CUInt
foreign import ccall safe "get_minimap2_data" c_get_minimap2_data :: CString

samtoolsData :: IO B.ByteString
samtoolsData =
    B.unsafePackCStringLen (c_get_samtools_data, convert c_get_samtools_len)

prodigalData :: IO B.ByteString
prodigalData =
    B.unsafePackCStringLen (c_get_prodigal_data, convert c_get_prodigal_len)

bwaData :: IO B.ByteString
bwaData =
    B.unsafePackCStringLen (c_get_bwa_data, convert c_get_bwa_len)

megahitData :: IO B.ByteString
megahitData =
    B.unsafePackCStringLen (c_get_megahit_data, convert c_get_megahit_len)

minimap2Data :: IO B.ByteString
minimap2Data =
    B.unsafePackCStringLen (c_get_minimap2_data, convert c_get_minimap2_len)


